multimc: hardcode jdk paths and use latest jdk by default

musfay opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
New Minecraft release requires `jdk16` while previous releases requires `jdk8`. MultiMC doesn't have a environment variable to declare java paths so I hardcoded those. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

musfay committed:
multimc: hardcode jdk paths and use latest jdk by default

kira-bruneau commented on pkgs/games/multimc/default.nix:
```diff
@@ -14,7 +13,13 @@ in mkDerivation rec {
-  buildInputs = [ qtbase jdk zlib ];
+  buildInputs = [ qtbase jdk8 zlib ];
+
+  postPatch = ''
```
Doesn't `JavaUtils.cpp` prioritize java located in `PATH` over everything else? https://github.com/MultiMC/MultiMC5/blob/develop/api/logic/java/JavaUtils.cpp#L353.

I think if you add `jdk16` to PATH like before, you shouldn't need this patch.

musfay replied:
If we add two `java` binaries to PATH, multimc(not only) will pick the first one. We should provide both versions of `java` as lots of people plays on older versions.

kira-bruneau replied:
`jre16` should be backwards compatible with `jre8`, so you should only need `jre16`. I'm assuming you're not using `jdk16` to build though since it fails to build with it?

starcraft66 replied:
I will have to test this to ensure it doesn't break things like modded forge installations. There are plenty of older minecraft configurations (especially modded ones) that rely on the hotspot jre version 8 from oracle.

starcraft66 commented:
I gave this a spin and it works mostly well, both java 8 and 16. However, I have identified a possible annoyance. MultiMC does not support autoselecting the correct java version for a Minecraft version. Furthermore, MultiMC appears to prefer java 8 over java 16 by default (possibly because it is the first in the list alphabetically?

<img width="327" alt="image" src="https://user-images.githubusercontent.com/1858154/122691328-f6f2c080-d1fc-11eb-91b5-3bb69d0e6bb8.png">

Now there are two ways to approach this. Java 8 could simply be the default and everything would mostly work out of the box except for Minecraft 1.17 (and its snapshots)+, however this could be confusing to people less "in the know" about java versions and minecraft versions. I think it might be more reasonable to set java 16 by default (this will likely require an extra patch on top of MultiMC because this doesn't seem configurable at all) which would allow Minecraft 1.17 to run as well as most older versions except whatever modded stuff breaks on Java >8. For those cases, more technical modding users could simply change the java version of their modded instances to Java 8.

What do you think @musfay?

musfay committed:
multimc: 0.6.12 -> unstable-2021-06-21

musfay commented:
> Now there are two ways to approach this. Java 8 could simply be the default and everything would mostly work out of the box except for Minecraft 1.17 (and its snapshots)+, however this could be confusing to people less "in the know" about java versions and minecraft versions. I think it might be more reasonable to set java 16 by default (this will likely require an extra patch on top of MultiMC because this doesn't seem configurable at all) which would allow Minecraft 1.17 to run as well as most older versions except whatever modded stuff breaks on Java >8. For those cases, more technical modding users could simply change the java version of their modded instances to Java 8.
> 

I prefer the second one as there is some weird forge instances which probably require Java 8. Updated multimc to latest commit and added patch to use latest java. They used to sort versions before listing on UI so I just reversed it.

SuperSandro2000 commented on pkgs/games/multimc/default.nix:
```diff
@@ -23,7 +34,7 @@ in mkDerivation rec {
 
     # xorg.xrandr needed for LWJGL [2.9.2, 3) https://github.com/LWJGL/lwjgl/issues/128
-    wrapProgram $out/bin/multimc --set GAME_LIBRARY_PATH /run/opengl-driver/lib:${libpath} --prefix PATH : ${jdk}/bin/:${xorg.xrandr}/bin/
+    wrapProgram $out/bin/multimc --set GAME_LIBRARY_PATH /run/opengl-driver/lib:${libpath} --prefix PATH : ${xorg.xrandr}/bin/
```
```suggestion
    wrapProgram $out/bin/multimc --set GAME_LIBRARY_PATH /run/opengl-driver/lib:${libpath} --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]}
```

musfay commented:
@starcraft66 Could you test this PR one more time? Also should we backport this PR as it adds support for Minecraft 1.17?

starcraft66 commented:
Yeah I want to test it yesterday but totally forgot. I will give it a go during my lunch time today.

As for backporting, I have no opposition. I run unstable but it would be nice for 1.17 to work on 21.05.

starcraft66 commented:
@musfay I just checked it out and the sorting of java versions doesn't seem quite right. MultiMC picks up the `java` executable I have installed globally in the list of JVMs and selects that by default instead of the highest java version number.

<img width="399" alt="image" src="https://user-images.githubusercontent.com/1858154/122979160-39cfa800-d365-11eb-8f88-be71f5bf2f74.png">

I took a look at the changes and it looks like the patch to fix the version sorting was removed when you force-pushed.

musfay commented:
@starcraft66 `multimc` picks the java from PATH and sets it as default before checking given paths. I just moved the caller function and It looks like it works as expected.
![Ekran görüntüsü_2021-06-22_21-59-49](https://user-images.githubusercontent.com/33374965/122984909-3a3c6300-d3a6-11eb-8ef2-3b98edd8476f.png)

starcraft66 reviewed (approved):
Thank you. LGTM
